server/utils/challenge/clients/challenge-client-v1-rsa.ts

```diff
@@ -0,0 +1,172 @@
+    signatureOrHashBuffer = await crypto.subtle.digest(
+      "SHA-256",
+      challengeRawDataBuffer
+    );
+  } else {
+    throw new Error("Cannot solve challenge.");
+  }
+
+  const signatureOrHashString = Buffer.from(signatureOrHashBuffer).toString(
+    "base64"
+  );
+
+  return `${CHALLENGE_CLIENT_VERSION}.${signatureOrHashString}` satisfies ChallengeSolutionWithVersion;
+}
+
+async function verifyChallenge({
+  // ChallengeData:
+  challenge,
+  session,
+  shareHash,
+
+  // Verification:
+  solution,
+  publicKey: publicKeyParam,
+}: VerifyChallengeParams) {
+  const solutionValue = solution.split(".")[1];
+
+  if (!solutionValue) {
+    return ErrorMessages.CHALLENGE_UNEXPECTED_ERROR;
+  }
+
+  if (
+    isAnonChallenge(challenge) ||
+    challenge.type === ChallengeType.SIGNATURE
+  ) {
+    // SIGNATURE-BASED CHALLENGE:
+
+    if (!publicKeyParam) {
+      return ErrorMessages.CHALLENGE_MISSING_PK;
+    }
+
+    const publicJWK: JsonWebKey = {
+      e: "AQAB",
+      ext: true,
+      kty: "RSA",
+      n: publicKeyParam,
+    };
+
+    const publicKey = await crypto.subtle.importKey(
+      "jwk",
+      publicJWK,
+      IMPORT_KEY_ALGORITHM,
+      true,
+      ["verify"]
+    );
+
+    const challengeRawData = await getChallengeRawData({
+      challenge,
+      session,
+      shareHash,
+    });
+
+    const challengeRawDataBuffer = Buffer.from(challengeRawData);
+
+    const isSignatureValid = crypto.subtle.verify(
+      SIGN_ALGORITHM,
+      publicKey,
+      Buffer.from(solutionValue, "base64"),
+      challengeRawDataBuffer
+    );
+
+    if (!isSignatureValid) return ErrorMessages.CHALLENGE_INVALID;
+  } else if (process.env.NODE_ENV === "development") {
+    // HASH-BASED CHALLENGES:
+
+    const expectedSolution = await solveChallenge({
+      challenge,
+      session,
+      shareHash,
+    });
+
+    // The length of the solution, excluding the version prefix, should be:
+    //
+    // - 44 characters (String.length) when using a hash.
+    // - 512 bytes (ArrayBuffer.byteLength) when using a signature from a 4096 modulusLength RSA.
+    // - 256 bytes (ArrayBuffer.byteLength) when using a signature from a 2048 modulusLength RSA.
+    // - ...
+    //
+    // However, there's no need to validate that, as we'll be comparing the whole value anyways.
+
+    // Constant-time comparison probably not needed (as we are comparing hashes), but anyway, there's a theatrical
+    // exploit if not doing so. See https://security.stackexchange.com/questions/209807/why-should-password-hash-verification-be-time-constant
+    if (!timingSafeEqual(Buffer.from(expectedSolution, "utf16le"), Buffer.from(solution, "utf16le"))) return ErrorMessages.CHALLENGE_INVALID;
+  } else {
+    return ErrorMessages.CHALLENGE_UNEXPECTED_ERROR;
+  }
+
+  return null;
+
+}
+
+export const ChallengeClientV1: ChallengeClient<JWKInterface> = {
+  version: CHALLENGE_CLIENT_VERSION,
+  ttlMs: 120000, // 120 seconds
```

Comment:
I've extended these because:

- Some users were still having issues with expiration, as you can see in the logs.
- With the EdDSA changes, the RSA-derived keys are only computed when needed, that is, if once the client receives the challenge from the backend, it turns out to be a v1 challenge; before, we would start deriving those keys before requesting the challenge. This means that now this process will take slightly longer (but while only affect activation once (then they migrate to EdDSA) and old recovery files).